sklearn/cluster/hierarchical.py

```diff
@@ -80,6 +80,56 @@ def _fix_connectivity(X, connectivity, affinity):
     return connectivity, n_components
 
 
+def _single_linkage_tree(connectivity, n_samples, n_nodes, n_clusters,
+                         n_components, return_distance):
+    """
+    Perform single linkage clustering on sparse data via the minimum
+    spanning tree from scipy.sparse.csgraph, then using union-find to label.
+    The parent array is then generated by walking through the tree.
+    """
+    from scipy.sparse.csgraph import minimum_spanning_tree
+
+    # explicitly cast connectivity to ensure safety
+    connectivity = connectivity.astype('float64')
```

Comment:
Would it be possible to use fused type and support both float64 and float32, in the interest of memory? We've been slowly but surely trying to make it so that in scikit-learn there is support of 32 and 64 bit in the interest of memory. It also tends to make code faster (data fits more in CPU cache).

Reply:
I would certainly be willing to look into that (I don't currently know how, but presume it can't be too hard). As @jakevdp pointed out, however, the first thing the ``minimum_spanning_tree`` will do is convert to ``float64`` so ultimately it is going to end up converted in a few lines anyway; by converting here we can do the finessing of zero distances without getting into type specific difficulties (in particular integer types). Is it worth using the fused type here and then letting scipy do the conversion in ``minimum_spanning_tree``?